pkg/policy/mapstate.go

```diff
@@ -860,28 +762,53 @@ func (ms *mapState) insertWithChanges(key Key, entry mapStateEntry, features pol
 // deleted.
 //
 // Incremental changes performed are recorded in 'changes'.
-func (ms *mapState) denyPreferredInsertWithChanges(newKey Key, newEntry mapStateEntry, features policyFeatures, changes ChangeState) {
-	// Bail if covered by a deny key
-	if features.contains(denyRules) {
+func (ms *mapState) insertWithChanges(newKey Key, newEntry mapStateEntry, features policyFeatures, changes ChangeState) {
+	if newEntry.IsDeny() {
+		// Bail if covered by a key of higher precedence, or by a deny key of the same
+		// precedence
 		for k, v := range ms.BroaderOrEqualKeys(newKey) {
-			// Identical deny key needs to be added to merge their entries.
-			if v.IsDeny && !(newEntry.IsDeny && k == newKey) {
+			if v.Precedence > newEntry.Precedence ||
+				// New deny entry is also bailed due to different covering deny key
+				// of the same precedence, equal keys need to be merged
+				v.Precedence == newEntry.Precedence && k != newKey {
 				return
 			}
 		}
-	}
 
-	if newEntry.IsDeny {
-		// Delete covered entries
+		// Delete covered entries of lower precedence, and
+		// same precedence deny entries if the keys are different
```

Comment:
What is the case when NarrowerOrEqualKeys could return a different key?

Reply:
It is typical for `NarrowerOrEqualKeys` to return different keys, that would be all the keys with narrower port (range). But it can also return a key that is the same, and in that case we need to merge the entries, but only if the precedence of the entries is the same. For example equal keys with same precedence originating from different rules/selectors, rule labels need to be merged. Could also be two proxy redirects on same precedence, here the tie breaker rule for proxy ports is used to consistently select one of them. Similar consideration for Auth requirement.